python/lsst/cp/pipe/bfCoeffCalc.py

```diff
@@ -0,0 +1,1272 @@
+                try:
+                    afwDisp.setDefaultBackend(self.debug.displayBackend)
+                    afwDisp.Display.delAllDisplays()
+                    self.disp1 = afwDisp.Display(0, open=True)
+                    self.disp2 = afwDisp.Display(1, open=True)
+
+                    im = afwImage.ImageF(1, 1)
+                    im.array[:] = [[1]]
+                    self.disp1.mtv(im)
+                    self.disp1.erase()
+                except NameError:
+                    self.debug.display = False
+                    self.log.warn('Failed to setup/connect to display! Debug display has been disabled')
+
+        plt.interactive(False)  # stop windows popping up when plotting. When headless, use 'agg' backend too
+        self.config.validateIsrConfig(self.log)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the BfTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitPairs", help="The list of visit pairs to use, as a list of tuples "
+                            "enclosed in quotes e.g. \"(123,456),(789,987),(654,321)\""
+                            "NB: must be comma-separated-tuples with no spaces, enclosed in quotes!")
+        parser.add_id_argument("--id", datasetType="bfKernelNew", ContainerClass=BfDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    @pipeBase.timeMethod
+    def run(self, dataRef, visitPairs):
+        """Run the brighter-fatter measurement task.
+
+        For a dataRef (which is each ccd here), and given a list of visit pairs, calulate the
+        brighter-fatter kernel for the ccd.
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the CCD for the visits to be fit.
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+        """
+        xcorrs = {}  # dict of lists keyed by either amp or CCD depending on level
+        means = {}
+        kernels = {}
+
+        # setup necessary objects
+        ccdNum = dataRef.dataId[self.config.ccdKey]
+        if self.config.level == 'CCD':
+            xcorrs = {ccdNum: []}
+            means = {ccdNum: []}
+        elif self.config.level == 'AMP':
+            # NB: don't use dataRef.get('raw_detector') due to composites
+            detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+            ampInfoCat = detector.getAmpInfoCatalog()
+            ampNames = [amp.getName() for amp in ampInfoCat]
+            xcorrs = {key: [] for key in ampNames}
+            means = {key: [] for key in ampNames}
+
+        # calculate or retrieve the gains
+        if self.config.doCalcGains:
+            self.log.info('Beginning gain estimation for CCD %s'%ccdNum)
```

Comment:
Mainly pushing back just because it seems like a reasonable amount of work given I've always done it this way, so might not nail it the first time, and as you said, harder to test, plus it will be virtually no saving, as there are very few messages, on code that will be run very infrequently. It will take many thousands of times longer to fix (probably literally billions actually) than it will save in runtime, so leaving unless you shout. Will put the spaces in though of course.

Reply:
That seems a very reasonable approach.